### ISSUE TITLE
Deprecate default drivers for some database 

### ISSUE THREAD

**Issue by CaselIT** [sqlalchemy/sqlalchemy#10680]
The current default drivers selection for some database is outdated or not really useful anymore.
An idea for 2.1 is to start raising a deprecation warning when a default drivers is selected.

Optionally the default behaviour is something that can be kept were it makes sense. In particular oracle and sqlite seem the most logical candidate for keeping the default.

An idea could be:

|Database|Current default|Behaviour|Alternative keeping default were useful|
|--------|----------|---------|-------|
|Mssql| pyodbc| deprecate suggesting `+pyodbc` to keep curren driver | keep as is since not much happened in mssql driver land|
|mysql||deprecate suggesting `+`|
|Mariadb||same as mysql||
|Oracle|cx_oracle| deprecate suggesting default it will change to oracledb | deprecate suggesting to use `+oracledb`|
|Postgresql|psycopg2|deprecate suggesting to use `+psycopg2` to keep current driver||
|Sqlite|pysqlite|deprecate suggesting `+pysqlite`|keep as is|

@zzzeek thoughts?

**Comment by CaselIT** [comment]
> I think we would have to just deprecate the concept of a "default" driver overall. we can't keep picking winners here.

I agree, but maybe we have a default only where it makes sense to have one. sqlite is the obvious case here. A case can be made also for oracle, since cx_oracle is legacy and oracledb is the only other driver there is.

> if we were to deprecate urls without a driver, im sure you can imagine, based on past URL changes, how the public reaction would be.

I agree, but that has to start at some point, so better sooner than later. Also as it was demonstrated by the previous url change no-one cares about deprecation warning, so the public reaction will be when we actually remove the thing.

My general idea was the following regarding when to make the change:
- if we end up removing the default everywhere, then no earlier than 2.3/2.4
- if we end up keeping only some default (oracle and sqlite), then the new default can be at changed in 2.2, but the removal of the other default would be the same as the the above case, so 2.3/2.4

**Comment by maurerle** [comment]
> I think we would have to just deprecate the concept of a "default" driver overall. we can't keep picking winners here.
>
> if we were to deprecate urls without a driver, im sure you can imagine, based on past URL changes, how the public 
reaction would be.

why should the default driver behavior be deprecated?
From a user perspective of an application it makes sense to just let users give a `postgresql://` uri without taking care about which driver is used.
Sure, the application dev should decide explicitly which driver should be used - but I see the current behavior as very useful.

Otherwise one would probably have to add things like

```python
if db_uri.startswith("postgresql://"):
  db_uri = db_uri.replace("postgresql://", "postgresql+psycopg2://")
```

in the application logic to keep the abstraction for users..?

**Comment by zzzeek** [comment]
psycopg2 itself is legacy by its author, as is cx_oracle, etc.   default drivers change externally to us.  this is why it's a bad idea

**Comment by zzzeek** [comment]
I dont have any appetite for doing anything here in 2.1.   I think what we're going to have to do is just hard-change the default for the backends where the driver has a successor.   In this case it would be psycopg3, oracledb.    I think this should be 2.2 for now

**Comment by CaselIT** [comment]
so you want to retain the default driver behavior long term?

**Comment by zzzeek** [comment]
for now yes.   not ready to have that fight

**Comment by CaselIT** [comment]
shall we warn in the 2.1 series regarding the change of default?
